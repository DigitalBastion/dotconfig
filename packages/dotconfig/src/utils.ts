import type { StandardSchemaV1 } from "@standard-schema/spec";
import type { IConfiguration, IConfigurationRoot, IConfigurationSection } from "./abstractions.js";
import { ConfigurationTypeSymbol } from "./constants.js";
import { ParseErrors } from "./errors.js";

/**
 * Get the iterator of key-value pairs for the configuration.
 * @param configuration The configuration to iterate.
 */
export function* configurationIterator(configuration: IConfiguration): Iterator<[string, string | null]> {
  const stack: IConfiguration[] = [configuration];

  while (stack.length > 0) {
    // biome-ignore lint/style/noNonNullAssertion: We already checking in the while loop.
    const config = stack.pop()!;

    if (config !== configuration && isConfigurationSection(config)) {
      yield [config.path, config.value];
    }
    for (const child of config.getChildren()) {
      stack.push(child);
    }
  }
}

/**
 * Parse the configuration using the schema.
 * @param schema The schema to use for parsing.
 * @param configuration The configuration to parse.
 */
export async function parseConfiguration<T extends StandardSchemaV1>(
  schema: T,
  configuration: IConfigurationSection,
): Promise<StandardSchemaV1.InferOutput<T>> {
  let result = schema["~standard"].validate(getConfigurationProxy(configuration));
  if (result instanceof Promise) {
    result = await result;
  }

  if (result.issues != null) {
    throw new ParseErrors(result.issues, { configurationPath: configuration.path });
  }

  return result.value;
}

function getConfigurationProxy(configuration: IConfiguration, parentPath: string | null = null): unknown {
  const configurationProxy = new Proxy(
    {},
    {
      get(_, getKey) {
        if (typeof getKey === "symbol" || getKey === "then") {
          return undefined;
        }

        const key = parentPath == null ? getKey : `${parentPath}:${getKey}`;
        if (configuration.getSection(key).getChildren().length > 0) {
          return getConfigurationProxy(configuration.getSection(key), key);
        }

        const value = configuration.get(getKey);
        return value;
      },
    },
  );

  return configurationProxy;
}

/**
 * Checks if the configuration is a section.
 * @param config The configuration to check.
 */
export function isConfigurationSection(config: IConfiguration): config is IConfigurationSection {
  return config[ConfigurationTypeSymbol] === "section";
}

/**
 * Checks if the configuration is a root.
 * @param config The configuration to check.
 */
export function isConfigurationRoot(config: IConfiguration): config is IConfigurationRoot {
  return config[ConfigurationTypeSymbol] === "root";
}
