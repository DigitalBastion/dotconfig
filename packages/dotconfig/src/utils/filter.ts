export function filterCaseInsensitiveDuplicates<T>(array: T[]): T[] {
    const set = new Set<string>();
    const result: T[] = [];

    for (const item of array) {
        const key = String(item).toLowerCase();

        if (set.has(key)) {
            continue;
        }

        set.add(key);
        result.push(item);
    }

    return result;
}
