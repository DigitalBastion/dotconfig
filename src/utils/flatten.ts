export class CircularReferenceError extends Error {
    constructor(public readonly path: string) {
        super("Circular references are not supported.");
    }
}

export type Primitive = string | number | boolean | null | undefined;
export type NestedObject = { [key: string]: unknown };

export interface FlattenOptions {
    delimiter?: string;
    transformKey?: (key: string) => string;
}

export function flatten(target: NestedObject, options: FlattenOptions = {}) {
    const delimiter = options.delimiter ?? '.';
    const transformKey = options.transformKey ?? ((key) => key);
    const result = new Map<string, Primitive>();
    const handledObjects = new WeakSet<NestedObject>();

    const stack: Array<{ obj: NestedObject; prefix: string }> = [{ obj: target, prefix: '' }];

    while (stack.length > 0) {
        const { obj, prefix } = stack.pop()!;

        for (const key of Object.keys(obj)) {
            const value = obj[key];
            const transformedKey = prefix !== "" ? prefix + delimiter + transformKey(key) : transformKey(key);

            if (typeof value !== "object" || value === null) {
                result.set(transformedKey, value as Primitive);
            } else if (Object.keys(value).length === 0) {
                continue;
            } else if (handledObjects.has(value as NestedObject)) {
                throw new CircularReferenceError(transformedKey);
            } else {
                stack.push({ obj: value as NestedObject, prefix: transformedKey });
                handledObjects.add(value as NestedObject);
            }
        }
    }

    return result;
}

function getKey(key: string): string | number {
    const parsedKey = Number(key);

    if (isNaN(parsedKey) || key.indexOf('.') !== -1) {
        return key;
    }

    return parsedKey;
}

function isEmpty(value: unknown): boolean {
    if (Array.isArray(value)) {
        return value.length === 0;
    } else if (typeof value === "object" && value !== null) {
        return Object.keys(value).length === 0;
    }

    return false;
}

export function unflatten(target: Map<string, unknown>, options: FlattenOptions = {}): NestedObject {
    const delimiter = options.delimiter ?? '.';
    const transformKey = options.transformKey ?? ((key) => key);
    const result: NestedObject = {};

    for (const [key, value] of target.entries()) {
        const keys = key.split(delimiter).map(transformKey);
        let currentObj = result;

        for (let i = 0; i < keys.length; i++) {
            const currentKey = getKey(keys[i]);

            if (i === keys.length - 1) {
                currentObj[currentKey] = value;
            } else {
                const nextKey = getKey(keys[i + 1]);

                if (!(currentKey in currentObj)) {
                    currentObj[currentKey] = typeof nextKey === 'number' ? [] : {};
                } else if (isEmpty(currentObj[currentKey])) {
                    currentObj[currentKey] = typeof nextKey === 'number' ? [] : {};
                }

                currentObj = currentObj[currentKey] as NestedObject;
            }
        }
    }

    return result;
}