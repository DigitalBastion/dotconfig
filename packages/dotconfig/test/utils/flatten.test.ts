import { expect, test } from "bun:test";
import { CircularReferenceError, flatten } from "../../src/utils.js";

test("Flattens empty nested keys.", () => {
  const input = {
    "": {
      "": {
        "": 1,
      },
    },
  };
  expect(flatten(input)).toMatchSnapshot();
});

test("Flattens a simple object.", () => {
  const input = { a: 1, b: { c: 2 }, d: [3, 4] };
  const expected = new Map([
    ["a", 1],
    ["b.c", 2],
    ["d.0", 3],
    ["d.1", 4],
  ]);
  expect(flatten(input)).toEqual(expected);
});

test("Handles nested objects.", () => {
  const input = { a: { b: { c: { d: 1 } } } };
  const expected = new Map([["a.b.c.d", 1]]);
  expect(flatten(input)).toEqual(expected);
});

test("Uses custom delimiter.", () => {
  const input = { a: { b: 1 }, c: { d: 2 } };
  const expected = new Map([
    ["a/b", 1],
    ["c/d", 2],
  ]);
  expect(flatten(input, { delimiter: "/" })).toEqual(expected);
});

test("Handles arrays.", () => {
  const input = { a: [1, 2, [3, 4]] };
  const expected = new Map([
    ["a.0", 1],
    ["a.1", 2],
    ["a.2.0", 3],
    ["a.2.1", 4],
  ]);
  expect(flatten(input)).toEqual(expected);
});

test("Handles null and undefined values.", () => {
  const input = { a: null, b: undefined, c: { d: null } };
  const expected = new Map([
    ["a", null],
    ["b", undefined],
    ["c.d", null],
  ]);
  expect(flatten(input)).toEqual(expected);
});

test("Uses custom key transformer.", () => {
  const input = { FOO: { BAR: 1 } };
  const expected = new Map([["foo_bar", 1]]);
  const transformKey = (key: string) => key.toLowerCase();
  expect(flatten(input, { transformKey, delimiter: "_" })).toEqual(expected);
});

test("Handles circular references.", () => {
  // biome-ignore lint/suspicious/noExplicitAny: Test case.
  const input = { a: 1 } as any;
  input.b = input;

  expect(() => flatten(input)).toThrow(new CircularReferenceError("b.b"));
});
