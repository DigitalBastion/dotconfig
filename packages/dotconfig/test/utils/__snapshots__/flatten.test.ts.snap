// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Flattens empty nested keys. 1`] = `
Map {
  ".." => 1,
}
`;
